cmd/kubectl/kubectlcobra/prune.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+// package kubectlcobra contains cobra commands from kubectl
+package kubectlcobra
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/kubectl/pkg/cmd/apply"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/validation"
+)
+
+// PruneOptions encapsulates the necessary information to
+// implement the prune functionality.
+type PruneOptions struct {
+	Client    dynamic.Interface
+	Builder   *resource.Builder
+	Mapper    meta.RESTMapper
+	Namespace string
+	// The currently applied objects (as Infos), including the
+	// current grouping object. These objects are used to
+	// calculate the prune set after retreiving the previous
+	// grouping objects.
+	CurrentGroupingObject *resource.Info
+	// The set of retrieved grouping objects (as Infos) selected
+	// by the grouping label. This set should also include the
+	// current grouping object. Stored here to make testing
+	// easier by manually setting the retrieved grouping infos.
+	PastGroupingObjects      []*resource.Info
+	RetrievedGroupingObjects bool
+
+	ToPrinter func(string) (printers.ResourcePrinter, error)
```

Comment:
I don't think we want to adopt this pattern from the ApplyOptions. It makes it hard to support more flexible UIs as well as providing this functionality as a library.

Reply:
Agree. But we should change the apply and prune printing together (in another PR).